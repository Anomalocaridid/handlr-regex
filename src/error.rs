use std::{path::PathBuf, process::ExitCode};

use tracing::{error, info};

/// Custom error type
#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error(transparent)]
    Io(#[from] std::io::Error),
    #[error(transparent)]
    Xdg(#[from] xdg::BaseDirectoriesError),
    #[error(transparent)]
    Config(#[from] confy::ConfyError),
    #[error("No handlers found for '{0}'")]
    NotFound(String),
    #[error(
        "Could not find a mimetype associated with the file extension: '{0}'"
    )]
    AmbiguousExtension(String),
    #[error(transparent)]
    BadMimeType(#[from] mime::FromStrError),
    #[error("Bad mime: {0}")]
    InvalidMime(mime::Mime),
    #[error("The desktop entry at '{0}' lacks a valid '{1}' field")]
    BadEntry(PathBuf, String),
    #[error("Error spawning selector process '{0}'")]
    Selector(String),
    #[error("Selection cancelled")]
    Cancelled,
    #[error("Please specify the default terminal with handlr set x-scheme-handler/terminal")]
    NoTerminal,
    #[error("Bad path: {0}")]
    BadPath(String),
    #[error(transparent)]
    SerdeJson(#[from] serde_json::Error),
    #[error(transparent)]
    SerdeIniDe(#[from] serde_ini::de::Error),
    #[error(transparent)]
    SerdeIniSer(#[from] serde_ini::ser::Error),
    #[error(transparent)]
    TracingGlobalDefault(#[from] tracing::dispatcher::SetGlobalDefaultError),
    #[error("Could not find file at path: '{0}'")]
    NonexistentFile(String),
    #[cfg(test)]
    #[error(transparent)]
    BadUrl(#[from] url::ParseError),
    #[cfg(test)]
    #[error(transparent)]
    FromUtf8(#[from] std::string::FromUtf8Error),
}

pub type Result<T, E = Error> = std::result::Result<T, E>;

#[mutants::skip] // Cannot completely test, relies on user input
pub fn handle(result: Result<()>) -> ExitCode {
    if let Err(error) = result {
        match error {
            // Cancelling the selector is an acceptable outcome
            Error::Cancelled => {
                info!("{}", error);
                ExitCode::SUCCESS
            }
            _ => {
                error!("{}", error);
                ExitCode::FAILURE
            }
        }
    } else {
        ExitCode::SUCCESS
    }
}
