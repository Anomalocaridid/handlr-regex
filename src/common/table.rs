use tabled::{
    settings::{themes::Colorization, Alignment, Color, Padding, Style},
    Table, Tabled,
};

/// Render a table from a vector of instances of Tabled structs
pub fn render_table<T: Tabled>(rows: &Vec<T>, terminal_output: bool) -> String {
    let mut table = Table::new(rows);

    if terminal_output {
        // If output is going to a terminal, print as a table
        table
            .with(Style::sharp())
            .with(Colorization::rows([Color::FG_WHITE, Color::BG_BLACK]))
    } else {
        // If output is being piped, print as tab-delimited text
        table
            .with(Style::empty().vertical('\t'))
            .with(Alignment::left())
            .with(Padding::zero())
    }
    .to_string()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::error::Result;
    use itertools::Itertools;

    #[derive(Tabled)]
    struct TestRow<'a> {
        col1: &'a str,
        col2: &'a str,
    }

    // Arbitrary sample text
    const LOREM_IPSUM: &str = "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.";

    // Helper function to create test data
    fn rows(test_text: &str) -> Vec<TestRow<'_>> {
        test_text
            .split(' ')
            .collect_vec()
            .chunks_exact(2)
            .map(|chunk| TestRow {
                col1: chunk[0],
                col2: chunk[1],
            })
            .collect_vec()
    }

    #[test]
    fn terminal_output() -> Result<()> {
        insta::assert_snapshot!(render_table(&rows(LOREM_IPSUM), true));
        Ok(())
    }

    #[test]
    fn piped_output() -> Result<()> {
        insta::assert_snapshot!(render_table(&rows(LOREM_IPSUM), false));
        Ok(())
    }
}
