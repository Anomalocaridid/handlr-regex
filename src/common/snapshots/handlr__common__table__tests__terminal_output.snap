---
source: src/common/table.rs
expression: "render_table(&rows(LOREM_IPSUM), true)"
---
┌──────────────┬───────────────┐
│[37m [39m[37mcol1[39m[37m        [39m[37m [39m│[37m [39m[37mcol2[39m[37m         [39m[37m [39m│
├──────────────┼───────────────┤
│[40m [49m[40mLorem[49m[40m       [49m[40m [49m│[40m [49m[40mipsum[49m[40m        [49m[40m [49m│
│[37m [39m[37mdolor[39m[37m       [39m[37m [39m│[37m [39m[37msit[39m[37m          [39m[37m [39m│
│[40m [49m[40mamet,[49m[40m       [49m[40m [49m│[40m [49m[40mconsectetur[49m[40m  [49m[40m [49m│
│[37m [39m[37madipiscing[39m[37m  [39m[37m [39m│[37m [39m[37melit,[39m[37m        [39m[37m [39m│
│[40m [49m[40msed[49m[40m         [49m[40m [49m│[40m [49m[40mdo[49m[40m           [49m[40m [49m│
│[37m [39m[37meiusmod[39m[37m     [39m[37m [39m│[37m [39m[37mtempor[39m[37m       [39m[37m [39m│
│[40m [49m[40mincididunt[49m[40m  [49m[40m [49m│[40m [49m[40mut[49m[40m           [49m[40m [49m│
│[37m [39m[37mlabore[39m[37m      [39m[37m [39m│[37m [39m[37met[39m[37m           [39m[37m [39m│
│[40m [49m[40mdolore[49m[40m      [49m[40m [49m│[40m [49m[40mmagna[49m[40m        [49m[40m [49m│
│[37m [39m[37maliqua.[39m[37m     [39m[37m [39m│[37m [39m[37mUt[39m[37m           [39m[37m [39m│
│[40m [49m[40menim[49m[40m        [49m[40m [49m│[40m [49m[40mad[49m[40m           [49m[40m [49m│
│[37m [39m[37mminim[39m[37m       [39m[37m [39m│[37m [39m[37mveniam,[39m[37m      [39m[37m [39m│
│[40m [49m[40mquis[49m[40m        [49m[40m [49m│[40m [49m[40mnostrud[49m[40m      [49m[40m [49m│
│[37m [39m[37mexercitation[39m[37m [39m│[37m [39m[37mullamco[39m[37m      [39m[37m [39m│
│[40m [49m[40mlaboris[49m[40m     [49m[40m [49m│[40m [49m[40mnisi[49m[40m         [49m[40m [49m│
│[37m [39m[37mut[39m[37m          [39m[37m [39m│[37m [39m[37maliquip[39m[37m      [39m[37m [39m│
│[40m [49m[40mex[49m[40m          [49m[40m [49m│[40m [49m[40mea[49m[40m           [49m[40m [49m│
│[37m [39m[37mcommodo[39m[37m     [39m[37m [39m│[37m [39m[37mconsequat.[39m[37m   [39m[37m [39m│
│[40m [49m[40mDuis[49m[40m        [49m[40m [49m│[40m [49m[40maute[49m[40m         [49m[40m [49m│
│[37m [39m[37mirure[39m[37m       [39m[37m [39m│[37m [39m[37mdolor[39m[37m        [39m[37m [39m│
│[40m [49m[40min[49m[40m          [49m[40m [49m│[40m [49m[40mreprehenderit[49m[40m [49m│
│[37m [39m[37min[39m[37m          [39m[37m [39m│[37m [39m[37mvoluptate[39m[37m    [39m[37m [39m│
│[40m [49m[40mvelit[49m[40m       [49m[40m [49m│[40m [49m[40messe[49m[40m         [49m[40m [49m│
│[37m [39m[37mcillum[39m[37m      [39m[37m [39m│[37m [39m[37mdolore[39m[37m       [39m[37m [39m│
│[40m [49m[40meu[49m[40m          [49m[40m [49m│[40m [49m[40mfugiat[49m[40m       [49m[40m [49m│
│[37m [39m[37mnulla[39m[37m       [39m[37m [39m│[37m [39m[37mpariatur.[39m[37m    [39m[37m [39m│
│[40m [49m[40mExcepteur[49m[40m   [49m[40m [49m│[40m [49m[40msint[49m[40m         [49m[40m [49m│
│[37m [39m[37moccaecat[39m[37m    [39m[37m [39m│[37m [39m[37mcupidatat[39m[37m    [39m[37m [39m│
│[40m [49m[40mnon[49m[40m         [49m[40m [49m│[40m [49m[40mproident,[49m[40m    [49m[40m [49m│
│[37m [39m[37msunt[39m[37m        [39m[37m [39m│[37m [39m[37min[39m[37m           [39m[37m [39m│
│[40m [49m[40mculpa[49m[40m       [49m[40m [49m│[40m [49m[40mqui[49m[40m          [49m[40m [49m│
│[37m [39m[37mofficia[39m[37m     [39m[37m [39m│[37m [39m[37mdeserunt[39m[37m     [39m[37m [39m│
│[40m [49m[40mmollit[49m[40m      [49m[40m [49m│[40m [49m[40manim[49m[40m         [49m[40m [49m│
│[37m [39m[37mid[39m[37m          [39m[37m [39m│[37m [39m[37mest[39m[37m          [39m[37m [39m│
└──────────────┴───────────────┘
